File: Numerals/ArabicToEthiopic.c
/* define number addresses */
#define ONE             0x1369
#define TEN             ONE + 9
#define HUNDRED         TEN + 9
#define TENTHOUSAND     HUNDRED + 1
typedef unsigned short  UCHAR;

#include <stdio.h>  /* imports "NULL */

 /**+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+/
/+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+/
//
//  arabtof (number)
//
//  Convert a string of ASCII digits into Ethiopic numerals.
//
//  number   -is a pointer to an ASCII string with a numeric sequence to be
//            convereted  ("1998" , "4,321" , ... )
//
//+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+/
/+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+**/


UCHAR*
ArabicToEthiopic (number)
  char* number;
{

UCHAR *ethioNumString, *returnString;
register int i, j, m, n;
char asciiOne, asciiTen;
UCHAR gezOne, gezTen, sep;
int place, pos;


  if ( number == NULL )
    return ( NULL );

  /*  simple integers only,
   *  reject if punctuation present:
   */
  for (i = 0; number[i] ; i++ )  
    if ( !(isdigit(number[i]) 
         || number[i] == '.'
         || number[i] == ',') )
      return ( NULL );

  n = strlen ( number ) - 1;

  if ( n == 0 ) {
    /* Only one digit, don't bother with the loop,
     * map onto Ethiopic "ones" and return:
     * */
    returnString = (UCHAR *) malloc ( 2 * sizeof (UCHAR) );
    returnString[0] = number[0] + ONE - '1';
    returnString[1] = '\0';
    return ( returnString );
  }
  else if ( (n % 2) == 0 ) {
     /* precondition the string to always have the leading
      * tens place populated, this avoids at least one check
      * in the loop:
      */
     char* tmp = (char *)malloc ( (n+2)*sizeof(char) );
     sprintf ( tmp, "0%s", number );
     number = tmp;
     n++;
  }


  /* make a big unicode digit holder: */
  ethioNumString = (UCHAR *) malloc ( (4*n+1) * sizeof (UCHAR) );

  m = 0;

  for ( place = n; place >= 0; place-- )
    {

      /* initialize values: */
      gezOne   = gezTen = 0x0;

      /* set ascii values: */
      asciiTen = number[n-place]; 
      place--;
      asciiOne = number[n-place]; 

      /* set ethiopic values: */
      if ( asciiOne != '0' )
        gezOne = asciiOne + (ONE - '1');  /* map onto Ethiopic "ones" */

      if ( asciiTen != '0' )
        gezTen = asciiTen + (TEN - '1');  /* map onto Ethiopic "tens" */

      /* pos indicates if our grouping subscript is even or odd       */
      pos = ( place % 4 ) / 2;

      /* find a separator, if any, to follow Ethiopic ten and one     */
      sep
      = ( place )
        ? ( pos )
          ? ( gezOne || gezTen )
            ? HUNDRED
            : 0x0
          : TENTHOUSAND
        : 0x0
      ;


      /* we want to clear ONE when it is superfluous  */
      if ( ( gezOne == ONE ) && gezTen == 0x0 )      /* one without a leading ten  */
        if ( ( sep == HUNDRED ) || (place+1) == n )  /* following (100) or leading */
          gezOne = 0x0;                              /* the sequence               */

      /* put it all together: */
      if ( gezTen )
        ethioNumString[m++] = gezTen;

      if ( gezOne )
        ethioNumString[m++] = gezOne;

      if ( sep )
        ethioNumString[m++] = sep;

    }

  returnString = ( UCHAR * ) malloc ( (m+1) * sizeof (UCHAR) );

  for (i=0; i<m; i++)
    returnString[i] = ethioNumString[i];

  returnString[m] = '\0';

  free ( ethioNumString );

  return ( returnString );

}

File: Numerals/ArabicToEthiopic.pl
#!/usr/bin/perl -w

binmode(STDIN, ":utf8");
binmode(STDOUT, ":utf8");
binmode(STDERR, ":utf8");
use utf8;
use strict;
use open ':encoding(utf-8)';


my @testNumbers = (
	"1",
	"10",
	"100",
	"1000",
	"10000",
	"100000",
	"1000000",
	"10000000",
	"100000000",
	"100010000",
	"100100000",
	"100200000",
	"100110000",
	"1",
	"11",
	"111",
	"1111",
	"11111",
	"111111",
	"1111111",
	"11111111",
	"111111111",
	"1111111111",
	"11111111111",
	"111111111111",
	"1111111111111",
	"1",
	"12",
	"123",
	"1234",
	"12345",
	"7654321",
	"17654321",
	"51615131",
	"15161513",
	"10101011",
	"101",
	"1001",
	"1010",
	"1011",
	"1100",
	"1101",
	"1111",
	"10001",
	"10010",
	"10100",
	"10101",
	"10110",
	"10111",
	"100001",
	"100010",
	"100011",
	"100100",
	"101010",
	"1000001",
	"1000101",
	"1000100",
	"1010000",
	"1010001",
	"1100001",
	"1010101",
	"101010101",
	"10000",          # Ilf
	"100000",         # aIlaf
	"1000000",        # aIlafat
	"10000000",       # tIlfit
	"100000000",      # tIlfitat
	"1000000000",
	"10000000000",
	"100000000000",   # mIlfit
	"1000000000000",  # mIlfitat
	"100010000",
	"100010100",
	"101010100",
	"3",
	"30",
	"33",
	"303",
	"3003",
	"3030",
	"3033",
	"3300",
	"3303",
	"3333",
	"30003",
	"30303",
	"300003",
	"303030",
	"3000003",
	"3000303",
	"3030003",
	"3300003",
	"3030303",
	"303030303",
	"333333333",
);



sub convert
{
my $number = $_[0];

	my $n = length ( $number ) - 1;

	unless ( $n % 2 ) {
		#
		#  Add dummy leading 0 to precondition the number for
		#  the algorithm and reduce one logic test within the
		#  for loop
		#
		$number = "0$number";
		$n++;
	}

	my @aNumberString = split ( //, $number );
	my $eNumberString = "";


	#
	#  read number from most to least significant digits:
	#
	for ( my $place = $n; $place >= 0; $place-- ) {
		#
		#  initialize values to emptiness:
		#
		my ($aTen, $aOne) = ( 0, 0);  #    ascii ten's and one's place
		my ($eTen, $eOne) = ('','');  # ethiopic ten's and one's place


		#
		#  populate our tens and ones places from the number string:
		#
		$aTen = $aNumberString[$n-$place]; $place--;
		$aOne = $aNumberString[$n-$place];


		#
		#  map to corresponding Ethiopic character addresses:
		#
		$eTen = chr( ord('፲') + ord($aTen) - ord('1') )  if( $aTen );
		$eOne = chr( ord('፩') + ord($aOne) - ord('1') )  if( $aOne );


		#
		#  pos tracks our 'pos'ition in a sequence of 4 digits
		#  to help determine what separator we need between
		#  a grouping of tens and ones.
		#
		my $pos = ( $place % 4 ) / 2;  # make even/odd 


		#
		#  find a separator, if any, to follow ethiopic ten and one:
		#
		my $sep
		= ( $place )
		  ? ( $pos ) # odd
		    ? ( ($eTen ne '') || ($eOne ne '') )
		      ? '፻'
		      : ''
		    : '፼'
		  : ''
		;


		#
		#  if $eOne is an Ethiopic '፩' we want to clear it under
		#  under special conditions.  These ellision rules could be
		#  combined into a single big test but gets harder to read
		#  and manage:
		#
		if ( ( $eOne eq '፩' ) && ( $eTen eq '' ) && ( $n > 1 ) ) {
			if ( $sep eq '፻' ) {
				#
				#  A superflous implied ፩ before ፻
				#
				$eOne = '';
			}
			elsif ( ($place+1) == $n ) {   # recover from initial $place--
				#
				#  ፩ is the leading digit.
				#
				$eOne = '';
			}
		}


		#
		#  put it all together and append to our output number:
		#
		$eNumberString .= "$eTen$eOne$sep";	
	}

	$eNumberString;
}



main:
{

	my $count = 1;;
	foreach ( @testNumbers ) {
		printf "%0i) ", $count++;	
		print "$_ => ", convert($_), "\n";

	}
	print "\n\n";
	print "See the Convert::Ethiopic::Lite::Number package at CPAN\n";

}
